### PR TITLE
Prevent uses of unsafe MySQL functions

### PR DESCRIPTION
**Pull request by jamiecobbett**
We replicate our data from master to slaves using MySQL's statement-based
replication. This means that SQL statements are relayed to the slave where it
runs them again. If a non-deterministic function call is part of the statement,
it will therefore save different data on the master and each slave. This
happened with the content IDs set in these migrations.

I removed the `randomized` scope from Classification as it failed the test but
hasn't been used since: 33a3b4c0976c83706c2191a29c69e0792470d558

Whilst we won't run these migrations again, I needed to change them so that the
test would pass and so that the unsafe examples aren't copied.

Setting each value in Ruby rather than single SQL statement will be much much
slower, especially for the 160,000 Document records, but we shouldn't ever run
these migrations again.

I based the code on TimeZoneTest.

### CODE REVIEW

File: db/data_migration/20141125162430_ensure_all_documents_have_content_id.rb [hunk 1]
```diff
@@ -1 +1,3 @@
-ActiveRecord::Base.connection.execute 'UPDATE documents SET content_id=UUID() where content_id IS NULL'
+Document.where(content_id: nil).each do |document|
```
Comment (alext) [inline db/data_migration/20141125162430_ensure_all_documents_have_content_id.rb]: Worth doing a `find_each` here to avoid loading them all into memory simultaneously?  (http://apidock.com/rails/ActiveRecord/Batches/find_each)

Comment (jamiecobbett) [inline db/data_migration/20141125162430_ensure_all_documents_have_content_id.rb]: I wondered about `find_in_batches` as a middle ground? (http://api.rubyonrails.org/classes/ActiveRecord/Batches.html#method-i-find_in_batches)

Comment (alext) [inline db/data_migration/20141125162430_ensure_all_documents_have_content_id.rb]: `find_each` is just a convenience wrapper around `find_in_batches`:
```ruby
    def find_each(options = {})
      if block_given?
        find_in_batches(options) do |records|
          records.each { |record| yield record }
        end
      else
        enum_for :find_each, options do
          options[:start] ? where(table[primary_key].gteq(options[:start])).size : size
        end
      end
    end
```

Comment (jamiecobbett) [inline db/data_migration/20141125162430_ensure_all_documents_have_content_id.rb]: Ha! Excellent, thanks @alext

**Comment by bradleywright** [comment]
:+1: for intent and adding tests (I haven't looked at the code though)

**Comment by bradleywright** [comment]
This looks good to me. Anyone object to merging?

**Comment by jabley** [comment]
:+1: Great test @jamiecobbett .

**Comment by jabley** [comment]
An observation – one thing that might bite is using the models directly in migrations, rather than defining a local version. I seem to recall a discussion with @floehopper about that in the distant past.

Appreciate it's a retrospective change, but part of the point of this PR is to discourage bad unsafe practices.

**Comment by heathd** [comment]
@jabley that's a good point and generally referencing models in rails database migrations is to be avoided. We specifically created data migrations to support those kind of scenarios where data is being manipulated and might want to use current active record model objects to support that scenario. So I think these changes are safe.

re the overall pull request I agree with Brad the intent is good, it's helpful to have automated reminders of these edge cases which otherwise can go by undetected.

I'm a bit concerned because mysql is case insensitive so `select uuid()` is equally as valid as `SELECT UUID()`. So this test might not actually achieve its goal. However if we made this test case insensitve then some of the checks (espcially `user()`) will cause false alarms.

Some possible compromises:

* make the checks case insensitive but only scan the `db/migrate` and `db/data_migration` directories.
* make the checks case insensitive, but only include functions which we are likely to use (ie avoid problematic functions `user()`, `rand()` and `sleep()`)

any thoughts?

**Comment by jamiecobbett** [comment]
@heathd you make a good point about false positives, especially with lowercase, which might match Ruby code. I've made it case insensitive, but it isn't catching any false positives. Maybe something to solve when it happens?

**Comment by heathd** [comment]
@jamiecobbett I think if we make it case insensitive and run across all ruby code, we should put a prominent comment in the code saying that the matching is pretty loose, and suggesting what to do if a false positive does occur.

**Comment by jamiecobbett** [comment]
@heathd good idea - comment added.

**Comment by heathd** [comment]
:+1: looks good
